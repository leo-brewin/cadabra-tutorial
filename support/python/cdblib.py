def create (file_name):
  import json, io, os, errno

  try:
      os.remove(file_name)                # delete the file if it exsists
      with open(file_name, 'w'): pass     # create an empty file
  except OSError as e:
      if e.errno == errno.ENOENT:         # errno.ENOENT = no such file or directory
         with open(file_name, 'w'): pass  # create an empty file
      else:
          raise                           # report an exception

  # Create and save an empty dict
  data_out = {}
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

# put and get based on ~/Python/examples/dict-io/lcb-01.py

def put (key_name,object,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_out = json.load(inp_file)

  # Add a new entry to the dict.

  # Two choices here. Both give almost identical output except for the rule symbol '->'.
  # Using str(object) will write this using unicode characters. This causes problems
  # down the track -- when these rules are imported in other codes we get a Python
  # error -- it reports that it can't handle this unicode character.
  # Using object.input_form() does the trick. It writes out '->'. This is good.

  # data_out[key_name] = str(object)       # will write rule char using unicode
  data_out[key_name] = object.input_form() # will write rule char as '->'

  # Save the updated dict
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

def get (key_name,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_inp = json.load(inp_file)

  # Return one entry from the dict
  return cadabra2.Ex (data_inp[key_name])
