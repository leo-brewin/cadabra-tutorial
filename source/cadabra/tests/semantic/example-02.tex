\documentclass[12pt]{cdblatex}
\usepackage{fancyhdr}
\usepackage{footer}

\begin{document}

% ============================================================================================
\section*{Example 2 Covriant derivatives}

\CdbSetup{action=hide}

\begin{cadabra}
   {a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u#}::Indices(position=independent).

   \nabla{#}::Derivative.
   \partial{#}::PartialDerivative.

   def check (objA,objB):

       tmp := @(objA) - @(objB).

       distribute     (tmp)
       sort_product   (tmp)
       rename_dummies (tmp)
       canonicalise   (tmp)

       return tmp

   import cdblib

   # ---------------------------------------------
   check000A = cdblib.get ('check000','expected/example-02.json')
   check001A = cdblib.get ('check001','expected/example-02.json')

   # ---------------------------------------------
   check000B = cdblib.get ('check000','output/example-02.json')
   check001B = cdblib.get ('check001','output/example-02.json')

   # ---------------------------------------------
   diff000 = check (check000A,check000B)   # cdb (ex-02.diff000,diff000)
   diff001 = check (check001A,check001B)   # cdb (ex-02.diff001,diff001)

\end{cadabra}

% \clearpage

\begin{dgroup*}
   \Dmath*{\cdb*{ex-02.diff000}}
   \Dmath*{\cdb*{ex-02.diff001}}
\end{dgroup*}

\end{document}
