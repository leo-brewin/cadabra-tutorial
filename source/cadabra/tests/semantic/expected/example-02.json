{
  "check000": "\\partial_{a}(v^{b}) + \\Gamma^{b c}_{a} v_{c}",
  "check001": "\\partial_{a}(v^{b}) + \\Gamma^{b}_{c a} v^{c}",
  "check002": "\\partial_{a}(u^{b}) + \\Gamma^{b}_{c a} u^{c} + \\partial_{a}(v^{b}) + \\Gamma^{b}_{c a} v^{c}"
}
